components/layout/query.rs

```diff
@@ -864,3 +877,167 @@ pub fn process_style_query<N: LayoutNode>(requested_node: N)
 
     StyleResponse(data.map(|d| d.styles.primary().clone()))
 }
+
+enum InnerTextItem {
+    Text(String),
+    RequiredLineBreakCount(u32),
+}
+
+// https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute
+pub fn process_element_inner_text_query<N: LayoutNode>(node: N,
+                                                       display_list: &Option<Arc<DisplayList>>) -> String {
+    if !display_list.is_some() {
+        warn!("We should have a display list at this point. Cannot get inner text");
+        return String::new();
+    }
+
+    // Step 1.
+    let mut results = Vec::new();
+    // Step 2.
+    inner_text_collection_steps(node, display_list.as_ref().unwrap(), &mut results);
+    let mut max_req_line_break_count = 0;
+    let mut inner_text = Vec::new();
+    for item in results {
+        match item {
+            InnerTextItem::Text(s) => {
+                if max_req_line_break_count > 0 {
+                    // Step 5.
+                    for _ in 0..max_req_line_break_count {
+                        inner_text.push("\u{000A}".to_owned());
+                    }
+                    max_req_line_break_count = 0;
+                }
+                // Step 3.
+                if !s.is_empty() {
+                    inner_text.push(s.to_owned());
+                }
+            },
+            InnerTextItem::RequiredLineBreakCount(count) => {
+                // Step 4.
+                if inner_text.len() == 0 {
+                    // Remove required line break count at the start.
+                    continue;
+                }
+                // Store the count if it's the max of this run,
+                // but it may be ignored if no text item is found afterwards,
+                // which means that these are consecutive line breaks at the end.
+                if count > max_req_line_break_count {
+                    max_req_line_break_count = count;
+                }
+            }
+        }
+    }
+    inner_text.into_iter().collect()
+}
+
+// https://html.spec.whatwg.org/multipage/#inner-text-collection-steps
+#[allow(unsafe_code)]
+fn inner_text_collection_steps<N: LayoutNode>(node: N,
+                                              display_list: &Arc<DisplayList>,
+                                              results: &mut Vec<InnerTextItem>) {
+    // Extracts the text nodes from the display list to avoid traversing it
+    // for each child node.
+    let mut text = HashMap::new();
+    for item in &display_list.as_ref().list {
+        if let &DisplayItem::Text(ref text_content) = item {
+            let entries = text.entry(&item.base().metadata.node).or_insert(Vec::new());
```

Comment:
The bad news is that #20031 conflicts with this, because it moves the text out of the display list.  The good news is that it should actually make it easier to query for the text of a node, because the text will already be in a HashMap, so you won't need to build one here...